pytest_plugins/issue_handlers.py

```diff
@@ -0,0 +1,294 @@
+        return handlers[handler_code.strip()](issue.strip(), data)
+
+
+def generate_issue_collection(items, config):  # pragma: no cover
+    """Generates a dictionary with the usage of Issue blockers
+
+    For use in pytest_collection_modifyitems hook
+
+    Arguments:
+        items {list} - List of pytest test case objects.
+        config {dict} - Pytest config object.
+
+    Returns:
+        [List of dicts] - Dicts indexed by "<handler>:<issue>"
+
+        Example of return data::
+
+            {
+                "XX:1625783" {
+                    "data": {
+                        # data taken from REST api,
+                        "status": ...,
+                        "resolution": ...,
+                        ...
+                        # Calculated data
+                        "is_open": bool,
+                        "is_deselected": bool,
+                        "clones": [list],
+                        "dupe_data": {dict}
+                    },
+                    "used_in" [
+                        {
+                            "filepath": "tests/foreman/ui/test_sync.py",
+                            "lineno": 124,
+                            "testcase": "test_positive_sync_custom_ostree_repo",
+                            "component": "Repositories",
+                            "usage": "skip_if_open"
+                        },
+                        ...
+                    ]
+                },
+                ...
+            }
+    """
+    settings.configure()
+    valid_markers = ["skip_if_open", "skip", "deselect"]
+    collected_data = defaultdict(lambda: {"data": {}, "used_in": []})
+
+    bz_cache_filename = config.getvalue('bz_cache')  # use existing json cache?
+    cached_data = None
+    if bz_cache_filename is not None:
+        try:
+            with open(bz_cache_filename, 'r') as bz_cache_file:
+                LOGGER.info(f'Using BZ cache file for issue collection: {bz_cache_filename}')
+                cached_data = {
+                    k: search_version_key(k, v) for k, v in json.load(bz_cache_file).items()
+                }
+        except FileNotFoundError:
+            # no bz cache file exists
+            LOGGER.warning(
+                f'bz-cache option used, file [{bz_cache_filename}] not found, '
+                f'NOT using cached data.'
+            )
+            pass
+
+    deselect_data = {}  # a local cache for deselected tests
+
+    test_modules = set()
+
+    # --- Build the issue marked usage collection ---
+    for item in items:
+        component = None
+        bzs = None
+        importance = None
+
+        # register test module as processed
+        test_modules.add(item.module)
+
+        # Find matches from docstrings top-down from: module, class, function.
+        mod_cls_fun = (item.module, getattr(item, 'cls', None), item.function)
+        for docstring in [d for d in map(inspect.getdoc, mod_cls_fun) if d is not None]:
```

Comment:
same falsey question for `is not None`

Reply:
ack - here I definitely want to be explicit, as its possible the docblock is an empty string (dumb, but possible?). I'm not nearly as certain that the return from `getdoc` will be `None`, like I would be with the pytest CLI `config.getvalue` call and check above.

Would shorten the composition a bit - optional or required?

Reply:
definitely optional